### PR TITLE
Implement solved vulnerability alerts

### PR DESCRIPTION
**Pull request by MiguelazoDS**
|Related issue|
|---|
|#20965|

## Description 

This PR implements the required changes to alert solved vulnerabilities by package removal

### Solved vulnerabilities

Logs 
image

Alerts
image

```json
{"timestamp":"2023-12-26T17:18:50.745-0300","rule":{"level":10,"description":"CVE-2023-1170 affects vim","id":"23505","firedtimes":48,"mail":false,"groups":["vulnerability-detector"],"gdpr":["IV_35.7.d"],"pci_dss":["11.2.1","11.2.3"],"tsc":["CC7.1","CC7.2"]},"agent":{"id":"001","name":"focal","ip":"192.168.33.20"},"manager":{"name":"jammy"},"id":"1703621930.1059644","decoder":{"name":"json"},"data":{"vulnerability":{"category":"Packages","classification":"CVSS","cve":"CVE-2023-1170","description":"Heap-based Buffer Overflow in GitHub repository vim/vim prior to 9.0.1376.","enumeration":"CVE","package":{"architecture":"amd64","description":"Vi IMproved - enhanced vi editor","install_time":" ","name":"vim","path":" ","size":"3048","type":"deb","version":"2:8.1.2269-1ubuntu5.21"},"reference":"https://huntr.dev/bounties/286e0090-e654-46d2-ac60-29f81799d0a4, https://github.com/vim/vim/commit/1c73b65229c25e3c1fd8824ba958f7cc4d604f9c, https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/DIAKPMKJ4OZ6NYRZJO7YWMNQL2BICLYV/, https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/IE44W6WMMREYCW3GJHPSYP7NK2VT5NY6/, https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/X4KDAU76Z7QNSPKZX2JAJ6O7KIEOXWTL/","score":{"base":"7.300000","version":"3.0"},"severity":"High","status":"Active"}},"location":"vulnerability-scanner"}
``` 

```json
{"timestamp":"2023-12-26T17:21:01.291-0300","rule":{"level":3,"description":"The CVE-2023-1170 that affected vim was solved due to a package removal/update or a system upgrade","id":"23502","firedtimes":153,"mail":false,"groups":["vulnerability-detector"],"gdpr":["IV_35.7.d"],"pci_dss":["11.2.1","11.2.3"],"tsc":["CC7.1","CC7.2"]},"agent":{"id":"001","name":"focal","ip":"192.168.33.20"},"manager":{"name":"jammy"},"id":"1703622061.1423515","decoder":{"name":"json"},"data":{"vulnerability":{"category":"Packages","classification":"CVSS","cve":"CVE-2023-1170","description":"Heap-based Buffer Overflow in GitHub repository vim/vim prior to 9.0.1376.","enumeration":"CVE","package":{"architecture":"amd64","name":"vim","version":"2:8.1.2269-1ubuntu5.21"},"reference":"https://huntr.dev/bounties/286e0090-e654-46d2-ac60-29f81799d0a4, https://github.com/vim/vim/commit/1c73b65229c25e3c1fd8824ba958f7cc4d604f9c, https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/DIAKPMKJ4OZ6NYRZJO7YWMNQL2BICLYV/, https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/IE44W6WMMREYCW3GJHPSYP7NK2VT5NY6/, https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/X4KDAU76Z7QNSPKZX2JAJ6O7KIEOXWTL/","score":{"base":"7.300000","version":"3.0"},"severity":"High","status":"Solved"}},"location":"vulnerability-scanner"}
```

### Update 12/27/23 integrity_clear alert

image
image

```json
{"timestamp":"2023-12-27T15:08:52.607-0300","rule":{"level":3,"description":"Vulnerabilities cleared","id":"23507","firedtimes":1,"mail":false,"groups":["vulnerability-detector"],"gdpr":["IV_35.7.d"],"pci_dss":["11.2.1","11.2.3"],"tsc":["CC7.1","CC7.2"]},"agent":{"id":"001","name":"focal","ip":"192.168.33.20"},"manager":{"name":"jammy"},"id":"1703700532.3195050","decoder":{"name":"json"},"data":{"vulnerability":{"category":"Packages","status":"Clear"}},"location":"vulnerability-scanner"}
```

### Update 01/03/24 

After the fix https://github.com/wazuh/wazuh/pull/21174 was merged, and some rebase commits added. 

After re-sync all detected vulnerabilities are reported.
image

After removing the Vim package and its dependencies.
image
image

Solved alert example 

```json
{"timestamp":"2024-01-03T12:13:59.009-0300","rule":{"level":3,"description":"The CVE-2023-0054 that affected vim was solved due to a package removal/update or a system upgrade","id":"23502","firedtimes":158,"mail":false,"groups":["vulnerability-detector"],"gdpr":["IV_35.7.d"],"pci_dss":["11.2.1","11.2.3"],"tsc":["CC7.1","CC7.2"]},"agent":{"id":"001","name":"focal","ip":"192.168.33.20"},"manager":{"name":"jammy"},"id":"1704294839.9023696","decoder":{"name":"json"},"data":{"vulnerability":{"category":"Packages","classification":"CVSS","cve":"CVE-2023-0054","description":"Out-of-bounds Write in GitHub repository vim/vim prior to 9.0.1145.","enumeration":"CVE","package":{"architecture":"amd64","name":"vim","version":"2:8.1.2269-1ubuntu5.21"},"reference":"https://huntr.dev/bounties/b289ee0f-fd16-4147-bd01-c6289c45e49d, https://github.com/vim/vim/commit/3ac1d97a1d9353490493d30088256360435f7731, http://seclists.org/fulldisclosure/2023/Mar/17, https://lists.debian.org/debian-lts-announce/2023/06/msg00015.html, https://security.gentoo.org/glsa/202305-16, https://support.apple.com/kb/HT213670","score":{"base":"7.300000","version":"3.0"},"severity":"High","status":"Solved"}},"location":"vulnerability-scanner"}
```

```json
{"timestamp":"2024-01-03T12:13:59.022-0300","rule":{"level":3,"description":"The CVE-2022-4292 that affected vim was solved due to a package removal/update or a system upgrade","id":"23502","firedtimes":159,"mail":false,"groups":["vulnerability-detector"],"gdpr":["IV_35.7.d"],"pci_dss":["11.2.1","11.2.3"],"tsc":["CC7.1","CC7.2"]},"agent":{"id":"001","name":"focal","ip":"192.168.33.20"},"manager":{"name":"jammy"},"id":"1704294839.9025545","decoder":{"name":"json"},"data":{"vulnerability":{"category":"Packages","classification":"CVSS","cve":"CVE-2022-4292","description":"Use After Free in GitHub repository vim/vim prior to 9.0.0882.","enumeration":"CVE","package":{"architecture":"amd64","name":"vim","version":"2:8.1.2269-1ubuntu5.21"},"reference":"https://huntr.dev/bounties/da3d4c47-e57a-451e-993d-9df0ed31f57b, https://github.com/vim/vim/commit/c3d27ada14acd02db357f2d16347acc22cb17e93, https://security.netapp.com/advisory/ntap-20230113-0005/, https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/WYC22GGZ6QA66HLNLHCTAJU265TT3O33/, https://security.gentoo.org/glsa/202305-16","score":{"base":"7.800000","version":"3.0"},"severity":"High","status":"Solved"}},"location":"vulnerability-scanner"}
```

```json
{"timestamp":"2024-01-03T12:13:59.034-0300","rule":{"level":3,"description":"The CVE-2022-4293 that affected vim was solved due to a package removal/update or a system upgrade","id":"23502","firedtimes":160,"mail":false,"groups":["vulnerability-detector"],"gdpr":["IV_35.7.d"],"pci_dss":["11.2.1","11.2.3"],"tsc":["CC7.1","CC7.2"]},"agent":{"id":"001","name":"focal","ip":"192.168.33.20"},"manager":{"name":"jammy"},"id":"1704294839.9027450","decoder":{"name":"json"},"data":{"vulnerability":{"category":"Packages","classification":"CVSS","cve":"CVE-2022-4293","description":"Floating Point Comparison with Incorrect Operator in GitHub repository vim/vim prior to 9.0.0804.","enumeration":"CVE","package":{"architecture":"amd64","name":"vim","version":"2:8.1.2269-1ubuntu5.21"},"reference":"https://huntr.dev/bounties/385a835f-6e33-4d00-acce-ac99f3939143, https://github.com/vim/vim/commit/cdef1cefa2a440911c727558562f83ed9b00e16b, https://security.gentoo.org/glsa/202305-16, https://security.netapp.com/advisory/ntap-20230203-0007/","score":{"base":"6.800000","version":"3.0"},"severity":"Medium","status":"Solved"}},"location":"vulnerability-scanner"}
```

- Compilation without warnings in every supported platform
  - [x] Linux
- [x] Source installation
- [x] Added unit tests (for new features)

### CODE REVIEW

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/scanContext.hpp [hunk 1]
```diff
@@ -36,16 +36,18 @@ enum class ScannerType
     Unknown = 0,
     PackageInsert = 1,
     PackageDelete = 2,
-    Os = 3,
-    HotfixInsert = 4,
-    HotfixDelete = 5,
+    PackagesClear = 3,
+    Os = 4,
+    HotfixInsert = 5,
+    HotfixDelete = 6,
```
Comment (GabrielEValenzuela) [inline src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/scanContext.hpp:42]: I think we'll have a conflict here, with the PR #21118

**Review by GabrielEValenzuela — CHANGES_REQUESTED** [review]
LGTM! The overall structure is clear, and the code is easy to read and understand.
### Readability:
The variable names are descriptive, and the comments provide helpful insights into the logic.
### Testing:
I've executed the provided tests, and I'm pleased to report that they produced the expected results.
<details><summary>Test report</summary>
#### Test requisites
- Install manager from this branch
- Connect an agent/s.

```bash
# Alert generated

2023/12/28 13:32:26 wazuh-modulesd:vulnerability-scanner[43853] sendReport.hpp:58 at handleRequest(): DEBUG: Vulnerability detected report: 1:[007] (df716d2bbb60) any->vulnerability-scanner:{"vulnerability":{"category":"Packages","classification":"CVSS","cve":"CVE-2023-40217","description":"An issue was discovered in Python before 3.8.18, 3.9.x before 3.9.18, 3.10.x before 3.10.13, and 3.11.x before 3.11.5. It primarily affects servers (such as HTTP servers) that use TLS client authentication. If a TLS server-side socket is created, receives data into the socket buffer, and then is closed quickly, there is a brief window where the SSLSocket instance will detect the socket as \"not connected\" and won't initiate a handshake, but buffered data will still be readable from the socket buffer. This data will not be authenticated if the server-side TLS peer is expecting client certificate authentication, and is indistinguishable from valid TLS stream data. Data is limited in size to the amount that will fit in the buffer. (The TLS connection cannot directly be used for data exfiltration because the vulnerable code path requires that the connection be closed on initialization of the SSLSocket.)","enumeration":"CVE","package":{"architecture":"amd64","description":"Interactive high-level object-oriented language (version 3.10)","install_time":" ","name":"python3.10","path":" ","size":633,"type":"deb","version":"3.10.12-1~22.04.3"},"reference":"https://mail.python.org/archives/list/security-announce@python.org/thread/PEPLII27KYHLF4AK3ZQGKYNCRERG4YXY/, https://www.python.org/dev/security/, https://lists.debian.org/debian-lts-announce/2023/09/msg00022.html, https://lists.debian.org/debian-lts-announce/2023/10/msg00017.html, https://security.netapp.com/advisory/ntap-20231006-0014/","score":{"base":5.3,"version":"3.1"},"severity":"Medium","status":"Active"}}
2023/12/28 13:32:27 wazuh-modulesd:vulnerability-scanner[43853] sendReport.hpp:58 at handleRequest(): DEBUG: Vulnerability detected report: 1:[007] (df716d2bbb60) any->vulnerability-scanner:{"vulnerability":{"category":"Packages","classification":"CVSS","cve":"CVE-2021-46848","description":"GNU Libtasn1 before 4.19.0 has an ETYPE_OK off-by-one array size check that affects asn1_encode_simple_der.","enumeration":"CVE","package":{"architecture":"amd64","description":"Manage ASN.1 structures (runtime)","install_time":" ","name":"libtasn1-6","path":" ","size":133,"type":"deb","version":"4.18.0-4build1"},"reference":"https://gitlab.com/gnutls/libtasn1/-/issues/32, https://bugs.gentoo.org/866237, https://lists.debian.org/debian-lts-announce/2023/01/msg00003.html, https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/AV4SHDJF2XLB4CUPTBPQQ6CLGZ5LKXPZ/, https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/ECM2ELTVRYV4BZ5L5GMIRQE27RFHPAQ6/, https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/OGO7XST4EIJGX4B2ITZCYSWM24534BSU/, https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/V5LWOGF7QRMNFRUCZY6TDYQJVFI6MOQ2/, https://gitlab.com/gnutls/libtasn1/-/commit/44a700d2051a666235748970c2df047ff207aeb5, https://security.netapp.com/advisory/ntap-20221118-0006/","score":{"base":9.1,"version":"3.1"},"severity":"Critical","status":"Active"}}
2023/12/28 13:37:38 wazuh-modulesd:vulnerability-scanner[43853] sendReport.hpp:58 at handleRequest(): DEBUG: Vulnerability detected report: 1:[004] (81a53081c758) any->vulnerability-scanner:{"vulnerability":{"category":"Packages","classification":"CVSS","cve":"CVE-2022-48303","description":"GNU Tar through 1.34 has a one-byte out-of-bounds read that results in use of uninitialized memory for a conditional jump. Exploitation to change the flow of control has not been demonstrated. The issue occurs in from_header in list.c via a V7 archive in which mtime has approximately 11 whitespace characters.","enumeration":"CVE","package":{"architecture":"amd64","description":"GNU version of the tar archiving utility","install_time":" ","name":"tar","path":" ","size":960,"type":"deb","version":"1.34+dfsg-1ubuntu0.1.22.04.2"},"reference":"https://savannah.gnu.org/bugs/?62387, https://lists.fedoraproject.org/archives/list/package-announce                          94409440658128sts.fedoraproject.org/message/CRY7VEL4AIG3GLIEVCTOXRZNSVYDYYUD/, https://lists.fedoraproject.org/archives/list/package-announce                                       1sts.fedoraproject.org/message/X5VQYCO52Z7GAVCLRYUITN7KXHLRZQS4/, https://savannah.gnu.org/patch/?10307","score":{"base":5.5,"version":"3.1"},"severity":"Medium","status":"Active"}}


# Alert solved
2023/12/28 13:32:27 wazuh-modulesd:vulnerability-scanner[43853] sendReport.hpp:58 at handleRequest(): DEBUG: Vulnerability detected report: 1:[007] (df716d2bbb60) any->vulnerability-scanner:{"vulnerability":{"category":"Packages","classification":"CVSS","cve":"CVE-2021-46848","description":"GNU Libtasn1 before 4.19.0 has an ETYPE_OK off-by-one array size check that affects asn1_encode_simple_der.","enumeration":"CVE","package":{"architecture":"amd64","description":"Manage ASN.1 structures (runtime)","install_time":" ","name":"libtasn1-6","path":" ","size":133,"type":"deb","version":"4.18.0-4build1"},"reference":"https://gitlab.com/gnutls/libtasn1/-/issues/32, https://bugs.gentoo.org/866237, https://lists.debian.org/debian-lts-announce/2023/01/msg00003.html, https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/AV4SHDJF2XLB4CUPTBPQQ6CLGZ5LKXPZ/, https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/ECM2ELTVRYV4BZ5L5GMIRQE27RFHPAQ6/, https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/OGO7XST4EIJGX4B2ITZCYSWM24534BSU/, https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/V5LWOGF7QRMNFRUCZY6TDYQJVFI6MOQ2/, https://gitlab.com/gnutls/libtasn1/-/commit/44a700d2051a666235748970c2df047ff207aeb5, https://security.netapp.com/advisory/ntap-20221118-0006/","score":{"base":9.1,"version":"3.1"},"severity":"Critical","status":"Active"}}
2023/12/28 13:37:38 wazuh-modulesd:vulnerability-scanner[43853] sendReport.hpp:58 at handleRequest(): DEBUG: Vulnerability detected report: 1:[004] (81a53081c758) any->vulnerability-scanner:{"vulnerability":{"category":"Packages","classification":"CVSS","cve":"CVE-2022-48303","description":"GNU Tar through 1.34 has a one-byte out-of-bounds read that results in use of uninitialized memory for a conditional jump. Exploitation to change the flow of control has not been demonstrated. The issue occurs in from_header in list.c via a V7 archive in which mtime has approximately 11 whitespace characters.","enumeration":"CVE","package":{"architecture":"amd64","description":"GNU version of the tar archiving utility","install_time":" ","name":"tar","path":" ","size":960,"type":"deb","version":"1.34+dfsg-1ubuntu0.1.22.04.2"},"reference":"https://savannah.gnu.org/bugs/?62387, https://lists.fedoraproject.org/archives/list/package-announce                          94409440658128sts.fedoraproject.org/message/CRY7VEL4AIG3GLIEVCTOXRZNSVYDYYUD/, https://lists.fedoraproject.org/archives/list/package-announce                                       1sts.fedoraproject.org/message/X5VQYCO52Z7GAVCLRYUITN7KXHLRZQS4/, https://savannah.gnu.org/patch/?10307","score":{"base":5.5,"version":"3.1"},"severity":"Medium","status":"Active"}}
root@pr-test:/var/ossec# grep "Vulnerability solved"  logs/ossec.log 
2023/12/28 13:37:39 wazuh-modulesd:vulnerability-scanner[43853] sendReport.hpp:58 at handleRequest(): DEBUG: Vulnerability solved report: 1:[004] (81a53081c758) any->vulnerability-scanner:{"vulnerability":{"category":"Packages","classification":"CVSS","cve":"CVE-2022-48303","description":"GNU Tar through 1.34 has a one-byte out-of-bounds read that results in use of uninitialized memory for a conditional jump. Exploitation to change the flow of control has not been demonstrated. The issue occurs in from_header in list.c via a V7 archive in which mtime has approximately 11 whitespace characters.","enumeration":"CVE","package":{"architecture":"amd64","name":"tar","version":"1.34+dfsg-1ubuntu0.1.22.04.1"},"reference":"https://savannah.gnu.org/bugs/?62387, https://lists.fedoraproject.org/archives/list/package-announce                          94409440658128sts.fedoraproject.org/message/CRY7VEL4AIG3GLIEVCTOXRZNSVYDYYUD/, https://lists.fedoraproject.org/archives/list/package-announce                                       1sts.fedoraproject.org/message/X5VQYCO52Z7GAVCLRYUITN7KXHLRZQS4/, https://savannah.gnu.org/patch/?10307","score":{"base":5.5,"version":"3.1"},"severity":"Medium","status":"Solved"}}
```
</details>

> [!IMPORTANT]  
> It could be a mismatch between versions due to an issue with vendors. Work in progress to solve it.

Great job  Migue!

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/scanContext.hpp (hunk 1 above)
Comment (MiguelazoDS) [reply src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/scanContext.hpp:42]: Hi @GabrielEValenzuela what if we have to do the same with hotfixes? I use PackagesClear to be more specific, so we can use something like HotfixesClear.

Comment (GabrielEValenzuela) [reply src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/scanContext.hpp:42]: Amazing idea! Could you do it on this PR?

Comment (MiguelazoDS) [reply src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/scanContext.hpp:42]: The scope is only for packages, we could add the hotfixes implementation later, or maybe I didn't understand the comment.

Comment (GabrielEValenzuela) [reply src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/scanContext.hpp:42]: Solved!

**Comment by Dwordcito** [comment]
To be fixed in https://github.com/wazuh/wazuh/pull/21144

**Pull request closed without merging**